onnxruntime/core/providers/cpu/nn/conv_integer.cc

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef _MSC_VER
+#pragma warning(disable : 4244)
+#pragma warning(disable : 4267)
+#endif
+
+#include "core/providers/cpu/nn/conv_integer.h"
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+#include "public/gemmlowp.h"
+
+namespace onnxruntime {
+namespace contrib {
+Status ConvInteger::Compute(OpKernelContext* context) const {
+  size_t num_inputs = OpKernel::Node().InputDefs().size();
+  const Tensor* X = context->Input<Tensor>(0);
+  const Tensor* W = context->Input<Tensor>(1);
+  int32_t input_offset = 0, filter_offset = 0;
+  if (num_inputs >= 3) {
+    const Tensor* X_Zero_Point = context->Input<Tensor>(2);
+    if (X_Zero_Point->Shape().NumDimensions() == 0 ||
+        (X_Zero_Point->Shape().NumDimensions() == 1 && X_Zero_Point->Shape().GetDims().size() == 1)) {
+      input_offset = static_cast<int32_t>(*(X_Zero_Point->Data<uint8_t>()));
+    } else {
+      //TODO: Add support for per-channel quantization.
+      return Status(common::ONNXRUNTIME, common::FAIL, "Non per-tensor quantization is not supported now.");
+    }
+  }
+  if (num_inputs >= 4) {
+    const Tensor* W_Zero_Point = context->Input<Tensor>(3);
+    if (W_Zero_Point->Shape().NumDimensions() == 0 ||
+        (W_Zero_Point->Shape().NumDimensions() == 1 && W_Zero_Point->Shape().GetDims().size() == 1)) {
+      filter_offset = static_cast<int32_t>(*(W_Zero_Point->Data<uint8_t>()));
+    } else {
+      //TODO: Add support for per-channel quantization.
+      return Status(common::ONNXRUNTIME, common::FAIL, "Non per-tensor quantization is not supported now.");
+    }
+  }
+
+  const int64_t N = X->Shape()[0];
+  const int64_t C = X->Shape()[1];
+  const int64_t M = W->Shape()[0];
+  ORT_RETURN_IF_ERROR(ValidateInputShape(X, W));
+
+  std::vector<int64_t> kernel_shape = ComputeKernelShape(W->Shape());
+
+  if (kernel_shape.size() + 2 != W->Shape().NumDimensions()) {
```

Comment:
Both these checks can be moved inside ComputeKernelShape

Reply:
will do the refactoring, a bad idea to copy it from conv.cc :). Thanks a lot!